server/src/main/java/io/druid/initialization/Initialization.java

```diff
@@ -171,199 +162,111 @@ static void clearLoadedModules()
-                u.getUserInfo(),
-                uri
-            );
-          }
-        }
-        remoteRepositories.add(r);
-      }
-      catch (URISyntaxException e) {
-        throw Throwables.propagate(e);
-      }
+    final File rootHadoopDependenciesDir = new File(extensionsConfig.getHadoopDependenciesDir());
+    if (rootHadoopDependenciesDir.exists() && !rootHadoopDependenciesDir.isDirectory()) {
+      throw new ISE("Root Hadoop dependencies directory [%s] is not a directory!?", rootHadoopDependenciesDir);
     }
-
-    if (log.isTraceEnabled() || log.isDebugEnabled()) {
-      return new DefaultTeslaAether(
-          config.getLocalRepository(),
-          remoteRepositories.toArray(new Repository[remoteRepositories.size()])
-      );
+    final File[] hadoopDependenciesToLoad = new File[hadoopDependencyCoordinates.size()];
+    int i = 0;
+    for (final String coordinate : hadoopDependencyCoordinates) {
+      final DefaultArtifact artifact = new DefaultArtifact(coordinate);
+      final File hadoopDependencyDir = new File(rootHadoopDependenciesDir, artifact.getArtifactId());
+      final File versionDir = new File(hadoopDependencyDir, artifact.getVersion());
+      // find the hadoop dependency with the version specified in coordinate
+      if (!hadoopDependencyDir.isDirectory() || !versionDir.isDirectory()) {
+        throw new ISE(
+            String.format("Hadoop dependency [%s] didn't exist!?", versionDir.getAbsolutePath())
+        );
+      }
+      hadoopDependenciesToLoad[i++] = versionDir;
     }
+    return hadoopDependenciesToLoad;
+  }
 
-    PrintStream oldOut = System.out;
-    try {
-      System.setOut(
-          new PrintStream(
-              new OutputStream()
-              {
-                @Override
-                public void write(int b) throws IOException
-                {
-
-                }
-
-                @Override
-                public void write(byte[] b) throws IOException
-                {
-
-                }
-
-                @Override
-                public void write(byte[] b, int off, int len) throws IOException
-                {
-
-                }
-              }
-              , false, StringUtils.UTF8_STRING
-          )
-      );
-      return new DefaultTeslaAether(
-          config.getLocalRepository(),
-          remoteRepositories.toArray(new Repository[remoteRepositories.size()])
-      );
-    }
-    catch (UnsupportedEncodingException e) {
-      // should never happen
-      throw new IllegalStateException(e);
-    }
-    finally {
-      System.setOut(oldOut);
+  /**
+   * @param extension The File instance of the extension we want to load
+   *
+   * @return a URLClassLoader that loads all the jars on which the extension is dependent
+   *
+   * @throws MalformedURLException
+   */
+  public static URLClassLoader getClassLoaderForExtension(File extension) throws MalformedURLException
+  {
+    URLClassLoader loader = loadersMap.get(extension.getName());
```

Comment:
Yeah, this is called from ```getFromExtensions(xxx)```, which is called twice, once in https://github.com/druid-io/druid/blob/master/services/src/main/java/io/druid/cli/Main.java#L76
and other in https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/initialization/Initialization.java#L408